rpc/core/env.go

```diff
@@ -35,13 +36,23 @@ const (
 
 var (
 	// set by Node
-	env *Environment
+	mut       = &sync.Mutex{}
```

Comment:
That's a super simple optimization todo that guarantees to work. 

> This PR bloats the diff between tendermint and celestia-core so we may seek alternatives even if this does work.

This is better be upstreamed OR if it's fixed in some 0.37 branch to downstream. This issue was fixed in 0.35 btw.

Reply:
tendmerint doesn't have a lot of thorough rpc tests, so this definitely needs more in depth testing. I'm kinda surprised its working at all without the use a RWMutex, and will likely cause quite a few issues imo. For instance, I don't think we're testing for the impact of multiple RPC requests at once.

The real fix requires not using a global var, and/or not spinning up multiple nodes in the same process. Any mutexes we add are purely feeble attempts to get a green checkmark in CI, not actually make it capable of handling multiple in process nodes safely.

Reply:
> This issue was fixed in 0.35 btw

Thanks for sharing! 0.35 moved the relevant file [env.go](https://github.com/tendermint/tendermint/blob/v0.35.9/internal/rpc/core/env.go). The 0.35 branch appears to contain a change that made `env` not a global variable but I can't find an isolated change for it based on [this history](https://github.com/tendermint/tendermint/commits/v0.35.9/internal/rpc/core/env.go). I can dig here if we want to try and cherry-pick an 0.35 change onto our fork. More importantly, this issue doesn't look fixed on main [env.go](https://github.com/tendermint/tendermint/blob/b42c439776811a6fcab8e22fb97fb44cee5701b3/rpc/core/env.go#L36-L39)

It sounds like we have two long-term approaches:

> the real fix requires not using a global var

seems like we should fix this upstream in Tendermint's main branch and optionally get it cherry-picked to the 0.34 release branch then apply it in celestia-core

> and/or not spinning up multiple nodes in the same process

seems like we need to fix this in Swamp tests

So we have two decisions to make:
1. Do we still want this short-term fix PR?
2. For the long-term fix, do we want to fix this upstream in Tendermint or downstream in Swamp tests?

Reply:
> > and/or not spinning up multiple nodes in the same process

> seems like we need to fix this in Swamp tests

Each Swamp test uses its own _one_ instance of App node, so when you run any such test on its own, the race does not appear. The problem starts when you run `go test ./...` because then all the tests are run within one runtime with shared memory. So the mentioned _fix_ would mean making the Swamp start only one instance, even with multiple tests being run via the global state, again 😅.

Reply:
Created https://github.com/celestiaorg/celestia-core/issues/882